### PR TITLE
Implement new spec api

### PR DESCRIPTION
**Pull request by fujunwei**
- Rename ModelBuilder to MLGraphBuilder, Model to MLGraph
- Remove old api Compilation and merge the compute capability to MLGraph
- Create MLGraph with MLContext and remove the concrete GraphBuilder from backend

endtoend test and lenet example can be passed on both DML and openvino backend.

This pull request is to fix https://github.com/webmachinelearning/webnn-native/issues/7, @huningxin PTAL, thanks.

### CODE REVIEW

File: src/webnn_native/GraphBuilder.cpp [hunk 1]
```diff
@@ -42,92 +43,105 @@
...
+                BUILD_ERROR_AND_CALLBACK("Failed to add output when building graph.");
             }
         }
-        if (GetContext()->ConsumedError(model->Finish())) {
-            return ModelBase::MakeError(this);
+        if (GetContext()->ConsumedError(graph->Finish())) {
+            BUILD_ERROR_AND_CALLBACK("Failed to finish building graph.");
         }
-        return model.Detach();
+        callback(MLBuildStatus_Success, reinterpret_cast<MLGraph>(graph.Detach()), nullptr,
```
Comment (huningxin) [inline src/webnn_native/GraphBuilder.cpp:143]: The build method essentially include two steps in previous implementation: 1) build the graph; 2) compile the graph. It probably should invoke `graph->Compile(callback)` here and let backends to compile the graph and invoke the callback. Backend implementation may need to do it asynchronously.

File: src/webnn_native/dml/GraphDML.cpp [hunk 2]
```diff
@@ -608,25 +622,84 @@ namespace webnn_native { namespace dml {
                 mOutputs[name] = ::dml::ActivationIdentity(reshape);
             }
         }
+
+        // FIXME(nhu): implement async
```
Comment (huningxin) [inline src/webnn_native/dml/GraphDML.cpp:626]: Move the following code into `Compile` method.

File: src/webnn_native/Graph.h [hunk 3]
```diff
@@ -62,10 +59,10 @@ namespace webnn_native {
         virtual MaybeError Finish();
 
       private:
-        ModelBase(ModelBuilderBase* modelBuilder, ObjectBase::ErrorTag tag);
-        virtual void CompileImpl(WebnnCompileCallback callback,
+        virtual void ComputeImpl(NamedInputsBase* inputs,
+                                 MLComputeCallback callback,
                                  void* userdata,
-                                 CompilationOptions const* options);
+                                 NamedOutputsBase* outputs) = 0;
```
Comment (huningxin) [inline src/webnn_native/Graph.h:65]: Probably we need to add a `CompileImpl` method and allow backends to implement. Please see my comments below.

File: src/webnn_native/openvino/GraphIE.cpp [hunk 4]
```diff
@@ -219,44 +257,71 @@ namespace webnn_native { namespace ie {
...
-    MaybeError Model::Finish() {
+    MaybeError Graph::Finish() {
         IEStatusCode code = IE(ie_model_finish)(mIeModel);
         DAWN_TRY(CheckStatusCode(code, "IE finish creating model"));
+
+        // We may leverage https://dawn-review.googlesource.com/c/dawn/+/36360 to
+        // implement async compilation as standle-alone component.
+        // Create compilation for IE backend.
+        code = IE(ie_create_compilation)(mIeModel, &mIeCompilation);
+        DAWN_TRY(CheckStatusCode(code, "IE create compilation"));
```
Comment (huningxin) [inline src/webnn_native/openvino/GraphIE.cpp:268]: ditto.

File: src/webnn_native/GraphBuilder.cpp (hunk 1 above)
Comment (fujunwei) [reply src/webnn_native/GraphBuilder.cpp:143]: Thanks, done.

**Comment by fujunwei** [comment]
@huningxin please take a look again, thanks.

File: src/webnn_native/GraphBuilder.cpp [hunk 5]
```diff
@@ -140,8 +140,10 @@ namespace webnn_native {
         if (GetContext()->ConsumedError(graph->Finish())) {
             BUILD_ERROR_AND_CALLBACK("Failed to finish building graph.");
         }
-        callback(MLBuildStatus_Success, reinterpret_cast<MLGraph>(graph.Detach()), nullptr,
-                 userdata);
+        // Compile the graph, call back the graph to user if sucessful, otherwise delete it in async
+        // thread like
+        // https://github.com/otcshare/webnn-native/commit/c16976d063b5dbbe9de85cd12a979814423fd232.
```
Comment (huningxin) [inline src/webnn_native/GraphBuilder.cpp:145]: Please show the code here, otcshare/webnn-native is a private repo.

Comment (fujunwei) [reply src/webnn_native/GraphBuilder.cpp]: Remove the private code.

File: src/webnn_native/GraphBuilder.cpp [hunk 6]
```diff
@@ -42,92 +43,106 @@
...
             }
         }
-        if (GetContext()->ConsumedError(model->Finish())) {
-            return ModelBase::MakeError(this);
+        if (GetContext()->ConsumedError(graph->Finish())) {
+            BUILD_ERROR_AND_CALLBACK("Failed to finish building graph.");
         }
-        return model.Detach();
+        // Compile the graph, call back the graph to user if sucessful, otherwise delete it in async
+        // thread.
```
Comment (huningxin) [inline src/webnn_native/GraphBuilder.cpp:144]: Could you please point me the code that deletes the graph if compile fails?

Comment (fujunwei) [reply src/webnn_native/GraphBuilder.cpp]: Add lamda function to delete failed graph in the call back.

**Review by huningxin — APPROVED** [review]
LGTM, thanks @fujunwei !

**Pull request merged**